operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/workflow/WorkflowReconcileExecutor.java

```diff
@@ -0,0 +1,293 @@
+package io.javaoperatorsdk.operator.processing.dependent.workflow;
+
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.javaoperatorsdk.operator.api.reconciler.Context;
+import io.javaoperatorsdk.operator.api.reconciler.dependent.Deleter;
+import io.javaoperatorsdk.operator.api.reconciler.dependent.DependentResource;
+import io.javaoperatorsdk.operator.api.reconciler.dependent.GarbageCollected;
+import io.javaoperatorsdk.operator.api.reconciler.dependent.ReconcileResult;
+
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class WorkflowReconcileExecutor<P extends HasMetadata> {
+
+  private static final Logger log = LoggerFactory.getLogger(WorkflowReconcileExecutor.class);
+
+  private final Workflow<P> workflow;
+
+  /** Covers both deleted and reconciled */
+  private final Set<DependentResourceNode> alreadyVisited = ConcurrentHashMap.newKeySet();
+  private final Set<DependentResourceNode> notReady = ConcurrentHashMap.newKeySet();
+  private final Map<DependentResourceNode, Future<?>> actualExecutions =
+      new HashMap<>();
+  private final Map<DependentResourceNode, Exception> exceptionsDuringExecution =
+      new ConcurrentHashMap<>();
+
+  private final Set<DependentResourceNode> markedForDelete = ConcurrentHashMap.newKeySet();
+  private final Set<DependentResourceNode> deleteConditionNotMet = ConcurrentHashMap.newKeySet();
+  // used to remember reconciled (not deleted or errored) dependents
+  private final Set<DependentResourceNode> reconciled = ConcurrentHashMap.newKeySet();
+  private final Map<DependentResource, ReconcileResult> reconcileResults =
+      new ConcurrentHashMap<>();
+
+  private final P primary;
+  private final Context<P> context;
+
+  public WorkflowReconcileExecutor(Workflow<P> workflow, P primary, Context<P> context) {
+    this.primary = primary;
+    this.context = context;
+    this.workflow = workflow;
+  }
+
+  public synchronized WorkflowExecutionResult reconcile() {
+    for (DependentResourceNode dependentResourceNode : workflow
+        .getTopLevelDependentResources()) {
+      handleReconcile(dependentResourceNode);
+    }
+    while (true) {
```

Comment:
Would be. First I tried with reactive, but this has quite dynamic nature, while with reactive (`CompletableFuture` etc) is meant to for static workflows. Pretty sure it would be possible somehow just probably not skilled enough in this.